script/hassfest/quality_scale_validation/test_before_setup.py

```diff
@@ -0,0 +1,66 @@
+"""Enforce that the integration raises correctly during initialisation.
+
+https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/test-before-setup/
+"""
+
+import ast
+
+from script.hassfest.model import Integration
+
+_VALID_EXCEPTIONS = {
+    "ConfigEntryNotReady",
+    "ConfigEntryAuthFailed",
+    "ConfigEntryError",
+}
+
+
+def _raises_exception(async_setup_entry_function: ast.AsyncFunctionDef) -> bool:
+    """Check that a valid exception is raised within `async_setup_entry`."""
+    for node in ast.walk(async_setup_entry_function):
+        if isinstance(node, ast.Raise):
+            if isinstance(node.exc, ast.Name) and node.exc.id in _VALID_EXCEPTIONS:
+                return True
+            if isinstance(node.exc, ast.Call) and node.exc.func.id in _VALID_EXCEPTIONS:
+                return True
+
+    return False
+
+
+def _calls_first_refresh(async_setup_entry_function: ast.AsyncFunctionDef) -> bool:
+    """Check that a async_config_entry_first_refresh within `async_setup_entry`."""
+    for node in ast.walk(async_setup_entry_function):
+        if (
+            isinstance(node, ast.Call)
+            and isinstance(node.func, ast.Attribute)
+            and node.func.attr == "async_config_entry_first_refresh"
```

Comment:
Some might raise `ConfigEntry...` errors from the coordinator update method and might not use `async_config_entry_first_refresh`. I haven't looked so maybe more a theoretical thing than someone actually doing it.

Reply:
The rule is about making sure it's called inside `async_setup_entry`
So if they raise `ConfigEntry...` in the coordinator update method but don't call `async_config_entry_first_refresh` then I think it's correct to fail the rule.